### ISSUE TITLE
Create the settings.php file if missing.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#1075]
Use case: wipe data from test installation in order to start over quickly...

1. drop/empty the db
2. delete ```./files``` and its contents
3. delete settings.php
4. visit site to re-start the installation process

As it is now, the installer complains about a missing settings.php:

![backdrop_installer-missing_settings php_file](https://cloud.githubusercontent.com/assets/2423362/8995117/d1ffab58-3752-11e5-9979-6fe45eedeaf4.png)

So I had to extract a fresh copy of the file in order to carry on. Can we ship a ```default-settings.php``` file and have the installer create a copy of it instead of complaining in this case?

**Comment by klonos** [comment]
...also the "place a copy of the file into..." implies a directory, but we point to a file. Would be a non-issue if we implemented this feature.

**Comment by docwilmot** [comment]
second this. would also solve the annoyance of accidental settings.php in Git commits.

**Comment by serundeputy** [comment]
>  delete settings.php

shouldn't you just not do this step?

**Comment by klonos** [comment]
>> delete settings.php

> just don't do this step?

Sure, there would be no issue with the ```$database``` or the ```$database_prefix``` or the ```$settings['hash_salt']``` entries in the old settings.php file, but it complains because it still references the (deleted from previous step) config folders (```$config_directories['active']``` & ```$config_directories['staging']```).

**Comment by klonos** [comment]
PS: I cannot test anything ATM because of #1072

**Comment by quicksketch** [comment]
I'd love to see settings.php trimmed down to the point where creating a new settings.php file is a simple as adding in the few settings that we use by default ($db_url, $config_directories, $settings['hash_salt']). However, then users would end up with a different settings.php that lacked all the common code comments (e.g. for setting up `$settings['reverse_proxy']` or `fast_404()`).

I'm not a big fan of the default-settings.php concept, but if we were to add such a file, would we just add it to the root of the installation? In D7 it was hidden under sties/default/. One of the reasons for moving it to the root was to reduce learning curve. I'd be afraid that having two settings.php files (even if one is called "default-settings.php") can cause confusion because its not clear that the default one isn't read in addition to the newly created one. Experienced Drupal users know that it's just there top copy, but new users can end up modifying the wrong one or not understanding that settings-default.php isn't executed at all.

**Comment by docwilmot** [comment]
I'm curious to know why you're not a fan, @quicksketch. What are the negatives apart from confusing new users?

And what if we hide `default-settings.php` in `core/misc` and copy it to root automatically when a user tries to install without one?

**Comment by quicksketch** [comment]
Hm, yeah stashing it away in core (anywhere, even just at /core/default-settings.php) would be okay. That's enough out of the way that the average user wouldn't poke around in there. The thing I would like to maintain however is that core comes with a settings.php in the root already, alleviating the requirement that users copy this file from default-settings.php, meaning we'd end up either with two copies of the file or need to create settings.php upon packaging. (Note that right now, we don't actually run the packager on core.)

**Comment by klonos** [comment]
> I'd love to see settings.php trimmed down to the point where creating a new settings.php file is a simple as adding in the few settings that we use by default ($db_url, $config_directories, $settings['hash_salt']).

How about if the auto-created (in Backdrop root by the installer) ```settings.php``` actually contained only those settings - no comments/documentation or anything else at all? Then, we could have ```default-settings.php``` (that could live inside ```core``` and perhaps also be just ```core/settings.php```) also be "executed", but any settings in root ```settings.php``` would override the ones in the default/core one? This way, we can ship any changes (and documentation) we want in ```core/settings.php``` without fear that we would mess with any settings the end user has in their setup.

In other words...

- lets introduce a ```settings.php``` that we "hide" inside ```core``` (that we already "train" users not to hack).
- we can update the core ```settings.php``` all we like with each release.
- lets auto-generate a second ```settings.php``` in root upon installation.
- The one inside core contains all documentation and any sensible defaults we think are suited for the 80%.
- we encourage people to read through the core ```settings.php``` but not edit it directly - copy changes to the root one instead.
- any setting inside the root ```settings.php``` overrides the equivalent setting inside the core one. Same as contrib modules and themes take precedence over core ones with the same name.

Far-fetched?

**Comment by sentaidigital** [comment]
I'm with @klonos on the idea of a default settings.php file, but I'm also with @quicksketch on the "two `settings.php` files will lead to endless issues" point.

I don't think we should have a `core/settings.php` or a `core/default-settings.php` file that is loaded by Backdrop.  If we need to set sensible defaults for anything, then that should be a function in `core/includes/bootstrap.inc` (or similar) or handled by CMI.  Blog posts would have to specify the root `settings.php` every time least new users start adding stuff in the copy in `core`.

Since the primary purpose is for documentation and to serve as a starting point, could we call it `settings.php.example` or `settings.php.default` or `settings.php.template` or anything that doesn't end in `.php` to make it clear that this is not in anyway something php should ever load, but is exactly the file you want if you're looking for a baseline `settings.php` file.  And maybe put it in `docs` or `core/docs` where PHP code doesn't go in the first place, but humans would if they're looking for documentation.

**Comment by klonos** [comment]
Well, D7 ships with a `/sites/default/default.settings.php` file that serves as a starting point if someone is forced for some reason to quickly wipe their settings and start over (which was the initial task I was trying to achieve before filing this issue).

**Comment by klonos** [comment]
...but I see your point in having the file not end in `.php`. If users know how to make a copy of `default.settings.php` and give it a different filename, then they can surely rename the extension. It's just that on most operating systems the default is to hide the file extension by default and thus we'd be forcing users to additional steps in order to enable visibility of file extensions in their file browser.

**Comment by olafgrabienski** [comment]
> It's just that on most operating systems the default is to hide the file extension by default

Good point! So what about `settings.php.example.txt`?

**Comment by sentaidigital** [comment]
Windows hides known extensions, and I don't think any app claims `.example` or `.default`, but I don't have a machine to test it.  Mac will hide _certain_ extensions like .jpg and .png, but left .pdf, .txt and .example alone.  (I just tried it on my laptop running 10.11.x.)

**Comment by mikemccaffrey** [comment]
I think we should stick with convention and name the file default.settings.php. That will be familiar to experienced drupal devs, while inexperienced ones shouldn't have any reason to be poking around in core/misc.

**Comment by docwilmot** [comment]
When creating multisites, the root `settings.php` already may have had changes applied and cannot be used for installing. I find I have to go hunt a freshly unpacked `backdrop.zip` and copy the `settings.php` from there. It would be useful if Backdrop could simply copy an existing `backup.settings.php` from core folder if there isn't one on install either in the root or in a sites folder.

**Comment by jlfranklin** [comment]
Currently, multisite installs will copy the `$BACKDROP_ROOT/settings.php` file if it is missing from the site-specific `sites/foo` directory.  Copying from `core/misc/` would be a much cleaner solution for the reasons @docwilmot just stated.

If we have any plans for a `core/docs` directory, that would be even better.  The `core/misc` has a bunch of live files, including CKEditor.  The default `settings.php` (under any name) would be an exception.
